### ISSUE TITLE
Allow selective inclusion of ETL backend libraries

### ISSUE THREAD

**Issue by metasim** [locationtech/geotrellis#2107]
To reduce the size of shaded jars and other `spark-submit` payloads, it would be helpfulI one could selectively include the backend libraries as needed. We tried marking `geotrellis-spark-etl` as `intransitive()` and then manually adding in the desired back-end (e.g. `geotrellis-accumulo`) , but the ETL framework eagerly instantiates the modules, causing `ClassNotFoundException`s on things relating to `hbase` and `cassandra`.

To de-couple the ETL from backend providers, I'd suggest using the Java SPI framework. It is very simple to use and implement for plugins conforming to a particular interface. Instead of having hard coded lists of I/O modules, you'd query the `ServiceLoader.load` method for a list of services discovered in the classpath. The official docs from Oracle are [here](https://docs.oracle.com/javase/tutorial/ext/basics/spi.html). Oddly, they make it look more complicated than it is in reality. You basically define an interface/trait for your API, and in your provider libraries add a special file in your resources declaring an implementation of the service.

cc: @mteldridge

**Comment by pomadchin** [comment]
Reasonable comments, thanks!
Related to #1767 
And it's definitely and API break.

**Comment by metasim** [comment]
@pomadchin Yes, we tried basically this:

```
libraryDependencies ++= Seq(
 geotrellis("raster"),
  geotrellis("spark"),
  geotrellis("spark-etl") intransitive(),
  "com.github.fge" % "json-schema-validator" % "2.2.6", // <-- needed for geotrellis-spark-etl
  geotrellis("accumulo")
)
```

but got `ClassNotFoundException` during the Spark job due to hbase, etc. GT classes. It appears that [this construction](https://github.com/locationtech/geotrellis/blob/master/spark-etl/src/main/scala/geotrellis/spark/etl/Etl.scala#L42) results in [this being called](https://github.com/locationtech/geotrellis/blob/master/spark-etl/src/main/scala/geotrellis/spark/etl/hbase/HBaseModule.scala#L21), which at some point leads to [this](https://github.com/locationtech/geotrellis/blob/master/spark-etl/src/main/scala/geotrellis/spark/etl/hbase/HBaseModule.scala#L21), which is in the `hbase` module. I'd have to get Matt to re-run to get the exact stack trace, so let me know if it's not clear.

**Comment by metasim** [comment]
Addressed by https://github.com/locationtech/geotrellis/pull/2223

@echeipesh Are there documentation updates planned? If not, should there be a ticket (along with some hints) that we can piggyback a WIP PR onto? As soon as 1.2 is released I'd like to update the `DataSource` reader you wrote for RasterFrames to use it, and need to learn how.

**Comment by echeipesh** [comment]
There should be a ticket to update the docs on this. Not totally sure if the rules of SemVer prevent us from changing the dependency structure here to target 1.2 (if the pain is acutely felt)

**Comment by metasim** [comment]
Ah, I missed the milestone tag.... so this is a 2.0 thing. I'll have to think about whether I want to wait for 2.0, or make a custom build etc.

**Comment by pomadchin** [comment]
@metasim btw we want to get rid of etl project at all and to replace it with https://github.com/locationtech/geotrellis/pull/2279

**Comment by echeipesh** [comment]
@metasim That was my question. Seems like the only thing we need to do is switch the subproject to `provided` and using the SPI interface in [`BackendProfiles`](https://github.com/locationtech/geotrellis/blob/master/spark-etl/src/main/scala/geotrellis/spark/etl/config/BackendProfiles.scala) to provide layerReader/Writer based on `LayerProvider` SPI then the user would have to include the individual io packages with the `--jars` tag.

Aside from changing dependencies to `provided` the other changes can be done w/o breaking the binary compatibility.

**Comment by pomadchin** [comment]
It's already implemented for a `pipeline` project https://github.com/locationtech/geotrellis/blob/master/build.sbt#L219
An ETL project is deprecated: 

The reason it still heavily depends on S3 as we don't have SPI for [S3GeoTiffRDD](https://github.com/locationtech/geotrellis/blob/master/s3/src/main/scala/geotrellis/spark/io/s3/S3GeoTiffRDD.scala)

**Comment by pomadchin** [comment]
It looks like `S3GeoTiffRDD` can be considered as a separate issue, @metasim feel free to create a new issue / reopen this one.
